README.md
**Overview**  
	This is my simple 6502 debugger and emulator. It takes plain binary images  
	for machine code.  
  
	Some things it supports yet:  
	* multiple CPUs  
	* some memory mapped IO  
		* Random Number Generator  
		* sebseb7's USB-SID-Module (broken ATM)  
		* Random Number Generator (broken)  
		* Keyboard Input (broken)  
  
**Benchmark**  
If you try it, PLEASE do a bench and mail the result to <nils.stec@gmail.com>.  
Please include information about your system. CPU, RAM and Kernel should be  
enough. If you run it on a "special" system (not PC, Server, ...) please   
write a bit more about your system. I will add these results to the following  
list.  
  
**How to build the emu and rum code**  
	$ make  
  
	$ ./6502 --rom bin_img/image_test1.65b 0x0000 --dbg  
		This loads image_test1.65b to address 0 and starts the debugger  
		see "debugger commands"  
  
	$ ./6502 --rom bin_img/image_test1.65b 0x0000 --bench 100000000  
		This will run the image for the given cycles and measure the   
		time the emu used. After that the imaginary speed of the 6502  
		cpu will be calculated.  
  
**Benchmark results**  
  
	* On a Intel Atom 1,6Ghz, 1024MB RAM i get about 90mhz...110mhz  
	  with the cpu core i use right now.   
	* Pentium 90/128MB, 2.6 => 5Mhz  
	* Pentium 4 3.2Ghz, 3.0 => 67Mhz  
	* Core2Duo 2.2Ghz, 8GB, 3.7 => 205Mhz  
	* i7 @ 1.2Ghz, 8GB, 3.2 => 430Mhz  
	* Samsung Galaxy S2, Android => 24Mhz  
  
  
**CLI arguments**  
	TBD  
	TBD  
	TBD  
  
**How to use it**  
  
	Debugger Commands:  
	d - dump mem  
	r - reset  
	i - irq  
	N - nmi  
	X - quit  
	v - show vectors  
	n - do a step  
	s - cpu status  
	S - set memory  
	p - set pc  
   
**The mysterious windows-tool "6502"**  
	IS NOT MINE. This is a example, what i want to have at the end.  
  
**Notes**  
	For this (still very simple) EMU we could use 2 different "CPU-cores".  
  
	At the moment we use a modified version of a 6502 cpu emulator, originally   
	written by <unknown>. It was used in Neil Bradley's Asteroids EMU and Steve  
	Green's MAC 6502 emu (google that, dont ask me!).  

	The first emu I used was a whole rewrite of this emu, but now it isn't working  
	anymore. I got a speed-increase by about 80%, depending on the machine code   
	that is executed, but atm it lacks some opcodes and others will throw   
	segfaults. It's working for the simple demo images i've made, but not more.  
	It is on the way, i think about 10% of the errors are fixed by now.   
  
	Before i started rewriting the core it ran a C64 rom-set and the REALLY COOL   
	demos from m6502 fine. Unfortunately the homepage of m6502 is down right now,  
	so we can't look at :(  

**changelog**  
		16th feb 2012 - please see file CHANGELOG from now on, i won't  
				update this changelog anymore.  
		25th jan 2012 - added memory mapped IO (random)  
				fixed some bugs  
				added opcode, nmi and irq hooks  
				  fixed some speed issues  

**LINKS**  
	PLEASE SEE http://www.6502asm.com/ - THE **VERY BEST** 6502 EMULATOR AROUND


CHANGELOG
09.03.2012:
* rewrite of makefile with a fancy help when you simply type make without
  further arguments
* working on another ncurses ui

17.02.2013:
* cleaned up a lot
* created a working-branch

16.02.2013:
>>> working again! <<<


begin of feb '13:
* changed the system to support multiple CPUs

/* the whole cpu and parts of the system are covered by this datatype.
 * this allows to create virtual systems with more than one cpu. a kind
 * of more or less SMP could be possible. i think of a RTOS running on a
 * fast ARM-chip.
 * we also hold the whole instruction set in this datatype, so we can
 * change opcodes on one and the other cpu remains unaffected. this can
 * be quite cool if we would emulate a 2-CPU-system with a 65c02 and a xxxxxx
 * with a little changed instruction set. 
 *
 * however, this is only the datatype for handling a cpu. the emu that
 * uses this untested black magic still has to be written.
 *
 * these things are implemented yet or will be soon:
 *
 *   + cpu (registers, irq/nmi/reset callback functions)
 *   + memory image
 *   + memory mapped I/O samples: simple rng, timer (not working right now)
 *   - frequency setting
 *
 *
 * of course, this needs a large amount of memory. at the moment each cpu 
 * consumes about 68k of ram.
 *    
 */

see 6502.h !


src/main.c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <time.h>
#include <string.h>
#include <termios.h>

#include "log.h"
#include "6502.h"
#include "main.h"
#include "nc_ui.h"
#include "ncui_new.h"

extern __6502_system_t cpu[N_CPUS];
//extern uint32_t active_cpu;

void run_image(uint32_t steps);
void debugger();
void print_help();

int32_t load_rom_image(char *filename, uint16_t offset);

uint16_t get_addr(char *str); 	// read 16bit hex from keyboard (ex: 3f00)
uint8_t get_hex8(char *str);
uint32_t get_dec32(char *str);
void term_block();		// keyboard settings, wait for return
void term_nonblock();		// keyboard settings, dont wait for return

uint16_t pc_init = 0;

#define ACTION_BENCH	1
#define ACTION_DEBUGGER	2
#define ACTION_NCURSES	3

void mem_clear() {
	int i;
	for(i = 0; i < 0x10000; i++) {
		ram[i] = 0;
	}
	_logf("memory: cleared");
	return;
}

int main(int argc, char **argv) {
	uint32_t i;
	uint16_t addr;

	int action = 0;
	uint32_t bench_steps = 0;

	_logf("--- started 6502 ---");

	mem_clear();


	if(argc <= 1) {
		_logf("starting ncurses ui");
		ncurses_ui();
		exit(0);
	}

	init_mmio();

	for(i = 1; i < argc; i++) {
		if(!(strcmp("-h", argv[i]))) {
			print_help();
			exit(0);
		}
		if(!(strcmp("--help", argv[i]))) {
			print_help();
			exit(0);
		}
		if(!(strcmp("--rom", argv[i]))) {
			addr = atoi(argv[i+2]);
			load_rom_image(argv[i+1],addr);
			i+=3;	
		}
		if(!(strcmp("--set-pc", argv[i]))) {
			pc_init = atoi(argv[i+1]);
			printf("initial pc=$%04x\n", pc_init);
			i+=2;
		}
		if(!(strcmp("--bench", argv[i]))) {
			//run_image(atoi(argv[i+1]));
			action = ACTION_BENCH;
			bench_steps = atoi(argv[i+1]);
		}
		if(!(strcmp("--dbg", argv[i]))) {
			//debugger();
			action = ACTION_DEBUGGER;
		}
		if(!(strcmp("--system-6502asm.com", argv[i]))) {

		}
		if(!(strcmp("--enable-random", argv[i]))) {
			uint16_t rng_base_addr = atoi(argv[i+1]);
			//rng_base_addr = 0xfe;
			printf("setting up 8bit RNG at address $%04x\n", rng_base_addr);
			i += 1;
			rng8_init(rng_base_addr, 0);
		}
		if(!(strcmp("--enable-keyboard", argv[i]))) {
			uint16_t kbd_base_addr = atoi(argv[i+1]);
			printf("setting up keyboard input at address $%04x\n", kbd_base_addr);
			i += 2;
		}
		if(!(strcmp("--cpus", argv[i]))) {
			uint32_t num_of_cpus = atoi(argv[i+1]);
			printf("setting up %d cpus\n", num_of_cpus);
			i += 2;
		}
		if(!(strcmp("--nc", argv[i]))||!(strcmp("--ncurses", argv[i]))) {
			action = ACTION_NCURSES;	
		}
	}

	switch(action) {
		case ACTION_BENCH:
			_logf("starting benchmark");
			run_image(bench_steps);
			break;
		case ACTION_DEBUGGER:
			_logf("starting debugger");
			printf("starting debugger\n");
			debugger();
			break;
		case ACTION_NCURSES:
			_logf("starting ncurses ui");
			ncurses_ui();
			break;
		default:
			print_help();
			exit(-1);
			break;
	}

	stop_mmio();

	return 0;
}


int32_t load_rom_image(char *filename, uint16_t offset) {
	uint32_t addr;
	uint8_t data;
	FILE *f = fopen(filename, "r");
	struct stat stbuf;

	stat(filename, &stbuf);

	_logf("load_rom_image(): file %s, size $%04x bytes", filename, (uint32_t)stbuf.st_size);

	if((stbuf.st_size + offset) > 0x10000) 
		_logf("load_rom_image(): warning! image exceeds memory, loading anyway, will be cut at 0xffff!");

	if(f == NULL) {
		_logf("load_rom_image(): WTF? can't open file...");
		exit(-2);
	}

	addr = offset;
	while(addr <= 0xffff) {
		data = getc(f);
		ram[addr] = data;
		addr++;	
	}

	fclose(f);
	_logf("load_rom_image(): rom image loaded");

	return 0;
}

void print_help() {
	printf("6502 emulator (c) 2010-2013 Nils Stec\n");
	printf(" --nc			start ncurses user interface (default when starting without arguments\n");
	printf(" --cpus <num>		how many CPUs to create/use\n");
	printf(" --rom <file> <offset>	load a plain binary image to <offset>\n");
	printf(" --bench	        do a benchmark\n");
	printf(" --dbg			start debugger\n");
	printf(" --set-pc <addr16>	set initial PC value\n");
	printf(" --system-6502asm.com	configure system to be compatible with 6502asm.com emu\n");
	printf(" --enable-random <addr16>	enable 8bit RNG at specific address\n");
	printf(" --enable-keyboard <addr16>	enable KBD input at specific address\n");
	printf(" --enable-video <mode>		enable Video output with specific mode: 0=c64 text, 1=6502asm.com\n");
	return;
}

void debugger() {
	int run = 1;
	int8_t kbuf = 0;
	uint16_t addr = 0;
	uint16_t rngaddr = 0;
	uint8_t val;

	printf("debugger\n");

	init6502();
	reset6502();

	term_nonblock();
	int retval;
	uint32_t steps = 0;
	uint32_t n_steps;
	uint32_t actcputmp;
	int x, y;
	kbuf = 's';

	cpu[get_cpu()].reg.pc = pc_init;

	while(run) {
		cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
		switch(kbuf) {
			case 'S':
				addr = get_addr("set memory\naddr");
				val = get_hex8("8bit hex");
				ram[addr] = val;
				goto cpustatus;
				break;
			case 'd':
				printf("memory dump\n");
				addr = get_addr("address");
				printf("       $.0 .1 .2 .3 .4 .5 .6 .7 .8 .9 .A .B .C .D .E .F\n");
				printf("       --- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --\n");
				for(y = 0; y < 16; y++) {
					printf(" $%04x: ", (unsigned int)addr);
					for(x = 0; x < 16; x++) {
						printf("%02x ", ram[addr]);
						addr++;
					}
					printf("\n");	
				}
				printf("\n");
				break;
			case 'D':
				printf("wide memory dump\n");
				addr = get_addr("address");
				printf("       $.0 .1 .2 .3 .4 .5 .6 .7 .8 .9 .A .B .C .D .E .F 10 11 12 13 14 15 16 17 18 19 1A 1B 1C 1D 1E 1F\n");
				for(y = 0; y < 32; y++) {
					printf(" $%04x: ", (unsigned int)addr);
					for(x = 0; x < 32; x++) {
						printf("%02x ", ram[addr]);
						addr++;
					}
					printf("\n");
				}
				printf("\n");
				break;
			case 'r':
				printf("reset\n");
				reset6502();
				cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				goto cpustatus;
				break;
			case 'i':
				printf("irq\n");
				irq6502();
#ifdef MMIO_USE_OPCODEHOOK
				mmio_irqhook();
#endif
				cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				goto cpustatus;
				break;
			case 'h':
				printf("avalaible commands\n\n");
				printf(" program execution:\n");
				printf("  n - do a step		p - set PC			G - couple of steps\n");
				printf("  r - reset		C - switch to another CPU\n");
				printf(" memory dumping/writing:\n");
				printf("  d - dump memory	D - dump memory (wide screen)	s - set memory\n");
				printf(" interrupts:\n");
				printf("  i - generate int	N - generate NMI");
				printf(" IO configuration:\n");
				printf("  R - set up RNG	V - set up video output		k - set up kbd\n");	 
				printf(" status:\n");
				printf("  v - show vectors	s - show CPU status		c - show sstem config\n");
				printf(" debugger control:\n");
				printf("  h - help		X - quit\n");
				goto cpustatus;
				break;
			case 'c':
				printf("system configuration:\n");
				printf(" * number of cpus: %d\n", get_cpus());
				break;
			case 'k':
				break;
			case 'R':
				rngaddr = get_addr("RNG base address, 0 for disable");
				if(rngaddr == 0) {
					rng8_deinit();
				} else {
					rng8_init(rngaddr, 0);
				}
				printf("RNG set to $%04x\n", rngaddr);
				break;
			case 'V':
				break;
			case 'C':
				actcputmp = get_cpu();
				char strtmp[50];
				sprintf(strtmp, "input number of CPU to use [%d]", actcputmp);
				set_cpu(get_dec32(strtmp));
				break;
			case 'p':
				cpu[get_cpu()].reg.pc = get_addr("set pc to");
				cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				goto cpustatus;
				break;
			case 'N':
				printf("nmi\n");
				cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				nmi6502();
#ifdef MMIO_USE_OPCODEHOOK
				mmio_nmihook();
#endif
				goto cpustatus;
				break;
			case 'X':
				printf("quit\n");
				run = 0;
				goto quit;
				break;
			case 'v':
				printf("vectors:\n");
				printf("NMI	$%02x%02x\n", ram[0xfffb], ram[0xfffa]);
				printf("RESET	$%02x%02x\n", ram[0xfffd], ram[0xfffc]);
				printf("IRQ/BRK $%02x%02X\n", ram[0xffff], ram[0xfffe]); 
				break;
			case 'n':
				rng8_getrnd();
				cpu[get_cpu()].reg.pc++;
				cpu[get_cpu()].inst.instruction[cpu[get_cpu()].opcode]();
				cpu[get_cpu()].ticks += cpu[get_cpu()].inst.opcode_ticks[cpu[get_cpu()].opcode];
				steps++;
#ifdef MMIO_USE_OPCODEHOOK
				mmio_opcodehook();
#endif
				cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				goto cpustatus;
				break;
			case 'G':
				n_steps = get_dec32("how many steps to run? ");
				for(;n_steps > 0; n_steps--) {
					rng8_getrnd();
					cpu[get_cpu()].reg.pc++;
					cpu[get_cpu()].inst.instruction[cpu[get_cpu()].opcode]();
					cpu[get_cpu()].ticks += cpu[get_cpu()].inst.opcode_ticks[cpu[get_cpu()].opcode];
					steps++;
#ifdef MMIO_USE_OPCODEHOOK
					mmio_opcodehook();
#endif	// MMIO_USE_OPCODEHOOK
					cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc];
				}
				goto cpustatus;
				break;
			case 's':
cpustatus:
				printf(" step %04d PC=$%04x dis:'%s", steps+1, cpu[get_cpu()].reg.pc, mnemonics[cpu[get_cpu()].opcode]);
				switch(opcode_len[cpu[get_cpu()].opcode]) {
					case 1:
						printf("        ");
						break;
					case 2:
						printf(" $%02x    ", ram[cpu[get_cpu()].reg.pc+1]);
						break;
					case 3:
						printf(" $%02x $%02x", ram[cpu[get_cpu()].reg.pc+1], ram[cpu[get_cpu()].reg.pc+2]);
						break;
				}
				printf("' opcode=$%02x(%db) A=$%02x X=$%02x Y=$%02x SP=$%02x S=$%02x cycles=%d cpu-cycles=%d\n", cpu[get_cpu()].opcode, opcode_len[cpu[get_cpu()].opcode], cpu[get_cpu()].reg.a, cpu[get_cpu()].reg.x, cpu[get_cpu()].reg.x, cpu[get_cpu()].reg.sp, cpu[get_cpu()].reg.flags, cpu[get_cpu()].inst.opcode_ticks[cpu[get_cpu()].opcode], cpu[get_cpu()].ticks);
				break;
		}
		retval = read(0, &kbuf, 1);
		retval++;
	}
quit:
	term_block();
}

void term_nonblock() {
	struct termios term = {0};

        tcgetattr(0, &term);    // err handling

        term.c_lflag &= ~ICANON;
        term.c_lflag &= ~ECHO;
        term.c_cc[VMIN] = 1;
        term.c_cc[VTIME] = 0;

        tcsetattr(0, TCSANOW, &term);   // err handling
	
	return;

}

void term_block() {
	struct termios term = {0};
	
	tcgetattr(0, &term);

	term.c_lflag |= ICANON;
	term.c_lflag |= ECHO;

	tcsetattr(0, TCSANOW, &term);

	return;
}

uint16_t get_addr(char *str) {
	unsigned int addr;
	
	term_block();

	printf("%s: ", str);
	if(!scanf("%x", &addr)) printf("error\n");

	term_nonblock();

	return (uint16_t)addr;
}

uint32_t get_dec32(char *str) {
	uint32_t val;

	term_block();

	printf("%s: ", str);
	if(!scanf("%d", &val)) printf("error\n");

	term_nonblock();

	return val;
}

uint8_t get_hex8(char *str) {
	unsigned int val;

	term_block();
	printf("%s: ", str);
	if(!scanf("%x", &val)) printf("error\n");

	term_nonblock();

	return (uint16_t)val;
}


void run_image(uint32_t steps) {
        uint32_t i;
	clock_t t1, t2;
        cpu[get_cpu()].reg.pc=0;

        init6502();
        printf("init 6502 ok\n");
        reset6502();
        printf("reset 6502 ok\n");

	printf("switching to cpu 0\n");
	set_cpu(0);

	printf("running %d steps...\n", steps);

        t1 = clock();
	for(i = 0; i < steps; i++) {
                cpu[get_cpu()].opcode = ram[cpu[get_cpu()].reg.pc++];
                cpu[get_cpu()].inst.instruction[cpu[get_cpu()].opcode]();
                cpu[get_cpu()].ticks += cpu[get_cpu()].inst.opcode_ticks[cpu[get_cpu()].opcode];
#ifdef MMIO_USE_OPCODEHOOK
	mmio_opcodehook();		
#endif
        }
	t2 = clock();

	float host_time = (float)(t2-t1)/CLOCKS_PER_SEC;
	float emu_speed_clk = (host_time*1000000)/cpu[get_cpu()].ticks;  
        float emu_speed_hz = 1000000.0/emu_speed_clk/1000000;

	printf(" 6502 ticks total: %d\n", cpu[get_cpu()].ticks);
	printf(" host time used:    %f\n", host_time);
	printf(" emu speed mhz:    %f\n", emu_speed_hz);
}

